src/tainting/Taint.ml

```diff
@@ -94,30 +75,129 @@ let _show_arg { pos = s, i; offset = os } =
     Printf.sprintf "arg(%s)#%d.%s" s i os_str
   else Printf.sprintf "arg(%s)#%d" s i
 
+(*****************************************************************************)
+(* Preconditions *)
+(*****************************************************************************)
+
+(* We roll our own boolean formula type here because we need to be
+   able to use polymorphic compare. Comparison on the Generic AST
+   isn't possible, and polymorphic compare would otherwise take into account
+   unimportant details like tokens.
+*)
+type precondition =
+  | Label of string
+  | Bool of bool
+  | And of precondition list
+  | Or of precondition list
+  | Not of precondition
+[@@deriving show]
+
+let rec expr_to_precondition e =
```

Comment:
I think it's better that we replace the `check in `parse_taint_requires` with a function  that already checks it _and_ convert it into `Taint.precondition` ? Maybe that requires moving the `precondition` type to `Rule`. But otherwise we would need to keep both in sync.

Reply:
you're so right. parse, don't validate